hy/importer.py

```diff
@@ -68,34 +71,77 @@ def import_file_to_ast(fpath, module_name):
     return hy_compile(import_file_to_hst(fpath), module_name)
 
 
-def import_file_to_module(module_name, fpath):
-    """Import content from fpath and puts it into a Python module.
+def import_file_to_module(module_name, fpath, loader=None):
+    """Import Hy source from fpath and put it into a Python module.
 
-    Returns the module."""
-    try:
-        _ast = import_file_to_ast(fpath, module_name)
-        mod = imp.new_module(module_name)
-        mod.__file__ = fpath
-        eval(ast_compile(_ast, fpath, "exec"), mod.__dict__)
-    except (HyTypeError, LexException) as e:
-        if e.source is None:
-            with open(fpath, 'rt') as fp:
-                e.source = fp.read()
-            e.filename = fpath
-        raise
-    except Exception:
-        sys.modules.pop(module_name, None)
-        raise
-    return mod
+    If there's an up-to-date byte-compiled version of this module, load that
+    instead. Otherwise, byte-compile the module once we're done loading it, if
+    we can.
 
+    Return the module."""
 
-def import_file_to_globals(env, module_name, fpath):
-    """ Import content from fpath and puts it into the dict provided
-    (e.g., for use in a REPL)
-    """
-    mod = import_file_to_module(module_name, fpath)
-    for k, v in mod.__dict__.items():
-        env[k] = v
+    module = None
+
+    bytecode_path = get_bytecode_path(fpath)
+    try:
+        source_mtime = int(os.stat(fpath).st_mtime)
+        with open(bytecode_path, 'rb') as bc_f:
+            # To get the bytecode file's internal timestamp, take the 4 bytes
+            # after the first 4 bytes and interpret them as a little-endian
+            # 32-bit integer.
+            bytecode_mtime = struct.unpack('<i', bc_f.read(8)[4:])[0]
+    except (IOError, OSError):
+        pass
+    else:
+        if bytecode_mtime >= source_mtime:
+            # It's a cache hit. Load the byte-compiled version.
+            if PY3:
+                # As of Python 3.6, imp.load_compiled still exists, but it's
+                # deprecated. So let's use SourcelessFileLoader instead.
+                from importlib.machinery import SourcelessFileLoader
+                module = (SourcelessFileLoader(module_name, bytecode_path).
+                          load_module(module_name))
+            else:
+                module = imp.load_compiled(module_name, bytecode_path)
```

Comment:
This part should be encased in a `try...except` and treat an `ImportError` as a cache miss. If you switch from one version of Python 2 to another, this will break due to a magic number mismatch, and you have to run `find . -name '*.pyc' -exec rm {} ';'` as a workaround.